src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java

```diff
@@ -0,0 +1,98 @@
+
+		converter = context.getLogicalToPhysicalPlanConverter();
+		PhysicalPlan optimalPlan = converter.convert(initialPlan, false);
+		
+		// Creating the completable future.
+		CostModel costModel = context.getCostModel();
+		CompletableFuture<Double> futureCost = costModel.initiateCostEstimation(optimalPlan);
+		
+		// Create rule instances.
+
+		// Get the cost.
+		Double localMinimum;
+		try {
+			localMinimum = futureCost.get();
+		} catch(final ExecutionException e) {
+			throw new QueryOptimizationException("CompletableFuture throws ExecutionException!", e);
+		} catch(final InterruptedException e) {
+			throw new QueryOptimizationException("CompletableFuture throws InterruptedException!", e);		
+		}
+		Double optimalCost, currentCost = localMinimum;
+		// currentCost = cost of the plan whose neighbours is being looked at.
+		// localMinimum = best cost among all neighbours
+		// optimalCost = best possible cost out of everything that the algorithm has found.
+		
+		// Optimization takes place here.
+		// Given the one starting state, find a local minimum.
+		// Firstly, generate neighbours. (Come to think of it, should I use the British or American spelling of neighbours?)
+		RewritingRule rwRule = null; // Eclipse suggested initiating this by adding the "= null". This doesn't seem quite right...
+		List<PhysicalPlan> neighbours = getNeighbours(optimalPlan,rwRule);
+		
+		List<CompletableFuture<Double>> futureNeighbourCosts = null; // Again, null suggested by Eclipse. Is this right?
+		List<Double> neighbourCosts;
+		
+		while(true) {
+			for(int i = 0; i < neighbours.size(); i++ ) {
+				// Here, I want to get an iterator for the first element of the list, and iterate through the list and get initiateCostEstimations for each one, filling up the futureNeighbourCosts
+			}
+			// Here, I want to have a second loop which gets the costs of the neighbours and compares them to the cost of the current plan looked at.
+			// If no neighbour beats the current plan, this part of the algorithm is over and the loop breaks.
+			// If any neighbour beats the current plan, the best neighbour becomes the new current plan.
+			
+			break; // Temporary break to not yield an error.
+		}
+		
+		return optimalPlan;
+	}
+	
+	/**
+	 * Currently created as a member method, but I might end up moving this to somewhere else if I find that it is needed in more places.
+	 * That obviously depends on whether Java supports importing methods outside of objects.
+	 * In C++, I would be passing references here - is something similar possible/appropriate in Java?
+	 * If significant performance is lost by making this a separate function, I'll be baking it into optimize. I've just been taught so far at LiU that making separate smaller functions is preferable so that the code becomes a bit more readable.
+	 * @return
+	 */
+	protected List<PhysicalPlan> getNeighbours(final PhysicalPlan initialPlan, final RewritingRule rewritingRule) {

```

Comment:
What is the purpose of this method? To get all neighbors that the given plan has in terms of the given rewriting rule? I thought that you actually want all possible neighbors, across all rewriting rules.

Reply:
My questions here are still open!

See also my earlier comment: https://github.com/LiUSemWeb/HeFQUIN/pull/115#issuecomment-951933055